parsers/ELEXON.py

```diff
@@ -0,0 +1,202 @@
+        hours=span[1]).format('YYYY-MM-DD')
+
+    if 'ELEXON_TOKEN' not in os.environ:
+        raise Exception('No ELEXON_TOKEN found! Please add it to secrets.env!')
+    params['APIKey'] = os.environ['ELEXON_TOKEN']
+    return session.get(ELEXON_ENDPOINT.format(report), params=params)
+
+
+def query_exchange(session, target_datetime=None):
+    params = {
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('INTERFUELHH', session, params, target_datetime)
+    response = response.content.decode()
+    return response
+
+
+def query_production(session, target_datetime=None):
+    params = {
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('FUELINST', session, params, target_datetime)
+    response = response.content.decode()
+    return response
+
+
+def parse_exchange(zone_key1, zone_key2, csv_text, target_datetime=None,
+                   logger=logging.getLogger(__name__)):
+    if not csv_text:
+        return None
+
+    sorted_zone_keys = sorted([zone_key1, zone_key2])
+
+    exchange = '->'.join(sorted_zone_keys)
+    data_points = list()
+    lines = csv_text.split('\n')
+
+    # check field count in report is as expected
+    field_count = len(lines[1].split(','))
+    if field_count != EXPECTED_FIELDS['INTERFUELHH']:
+        logger.warning(
+            'Expected {} fields in INTERFUELHH report, got {}'.format(
+                EXPECTED_FIELDS['INTERFUELHH'], field_count
+            )
+        )
+
+    for line in lines[1:-1]:
+        fields = line.split(',')
+
+        # create the datetime
+        datetime = datetime_from_date_sp(fields[1], fields[2])
+
+        data = {
+            'sortedZoneKeys': exchange,
+            'datetime': datetime,
+            'source': 'bmreports.com'
+        }
+
+        # positive value implies import to GB
+        multiplier = -1 if 'GB' in sorted_zone_keys[0] else 1
+        data['netFlow'] = float(fields[EXCHANGES[exchange]]) * multiplier
+        data_points.append(data)
+
+    if not target_datetime:
+        return data_points[-1]
+    else:
+        return data_points
+
+
+def parse_production(csv_text, target_datetime=None,
+                     logger=logging.getLogger(__name__)):
+    if not csv_text:
+        return None
+
+    data_points = list()
+    lines = csv_text.split('\n')
+
+    # check field count in report is as expected
+    field_count = len(lines[1].split(','))
+    if field_count != EXPECTED_FIELDS['FUELINST']:
+        logger.warning(
+            'Expected {} fields in FUELINST report, got {}'.format(
+                EXPECTED_FIELDS['FUELINST'], field_count
+            )
+        )
+
+    for line in lines[1:-1]:
+        fields = line.split(',')
+
+        # create the datetime
+        datetime = datetime_from_date_sp(fields[1], fields[2])
+
+        data = {
+            'zoneKey': 'GB',
+            'datetime': datetime,
+            'production': {},
+            'storage': {},
```

Comment:
can you confirm that positive storage values means using electricity from the grid, and that negative values means giving back (therefore acting as production)?
It's the opposite of the production logic.

Reply:
I will need to check that. Actually I might have implemented hydro wrong, ELEXON provides values for hydro pumped storage and hydro non-pumped storage. At the moment I have using pumped storage for storage and non-pumped storage for production. I'm not fully sure what the terms mean.

Reply:
We should put non-pumped hydro in production and pumped hydro in storage, but multiplied by -1. We should double check that pumped hydro can be both positive and negative in the data

Reply:
@lorrieq @corradio UK-GB previously hasn't been reporting to ENTSO-E what is stored into "pumped storage". 
ELEXON also looks like it does not contain this data. The PS-value never goes below 0. So there is nothing that shows the consumption for pumping water to the storage (= hydro storage bar does not go to negative side on the map).

So "pumped storage" reported by ELEXON is the *output* of electricity from discharging pumped storage. Positive values on ELEXON represent the output for the "hydro storage" category.

"Non-pumped storage" means, there either is a "reservoir type" hydro power with a natural inflow of water, which is released when it's needed, or a run-of-river hydro plant, which is using the recent flow of water of rivers.
Non-PS belongs to the usual "hydro" category.